service/frontend/adminHandler.go

```diff
@@ -421,6 +423,237 @@ func (adh *AdminHandler) GetWorkflowExecutionRawHistory(
+				VersionHistory: versionHistory.ToThrift(),
+			}, nil
+		}
+		return nil, err
+	}
+
+	token.PersistenceToken = rawHistoryResponse.NextPageToken
+	size := rawHistoryResponse.Size
+	// N.B. - Dual emit is required here so that we can see aggregate timer stats across all
+	// domains along with the individual domains stats
+	adh.metricsClient.RecordTimer(scope, metrics.HistorySize, time.Duration(size))
+	domainScope := adh.metricsClient.Scope(scope, metrics.DomainTag(request.GetDomain()))
+	domainScope.RecordTimer(metrics.HistorySize, time.Duration(size))
+
+	rawBlobs := rawHistoryResponse.HistoryEventBlobs
+	blobs := []*gen.DataBlob{}
+	for _, blob := range rawBlobs {
+		blobs = append(blobs, blob.ToThrift())
+	}
+
+	result := &admin.GetWorkflowExecutionRawHistoryV2Response{
+		HistoryBatches: blobs,
+		VersionHistory: versionHistory.ToThrift(),
+	}
+	if len(token.PersistenceToken) == 0 {
+		result.NextPageToken = nil
+	} else {
+		result.NextPageToken, err = serializeHistoryToken(token)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return result, nil
+}
+
+func (adh *AdminHandler) validateGetWorkflowExecutionRawHistoryV2Request(
+	request *admin.GetWorkflowExecutionRawHistoryV2Request,
+) error {
+
+	execution := request.Execution
+	if len(execution.GetWorkflowId()) == 0 {
+		return &gen.BadRequestError{Message: "Invalid WorkflowID."}
+	}
+	// TODO currently, this API is only going to be used by re-send history events
+	// to remote cluster if kafka is lossy again, in the future, this API can be used
+	// by CLI and client, then empty runID (meaning the current workflow) should be allowed
+	if len(execution.GetRunId()) == 0 || uuid.Parse(execution.GetRunId()) == nil {
+		return &gen.BadRequestError{Message: "Invalid RunID."}
+	}
+
+	pageSize := int(request.GetMaximumPageSize())
+	if pageSize < 0 {
+		return &gen.BadRequestError{Message: "Invalid PageSize."}
+	}
+	return nil
+}
+
+func (adh *AdminHandler) getEventRange(
+	request *admin.GetWorkflowExecutionRawHistoryV2Request,
+	versionHistories *persistence.VersionHistories,
+) (int64, int64, *persistence.VersionHistory, error) {
+
+	currentBranch, err := versionHistories.GetCurrentVersionHistory()
+	// The start event should be exclusive
+	targetStartEventID := request.GetStartEventId() + 1
+	targetStartEventVersion := request.GetStartEventVersion()
+	if request.StartEventId == nil || request.StartEventVersion == nil {
+		firstItem, err := currentBranch.GetFirstItem()
+		if err != nil {
+			return 0, 0, nil, err
+		}
+		targetStartEventID = common.FirstEventID
+		targetStartEventVersion = firstItem.GetVersion()
+	}
+	targetEndEventID := request.GetEndEventId()
+	targetEndEventVersion := request.GetEndEventVersion()
+	if request.EndEventId == nil || request.EndEventVersion == nil {
+		lastItem, err := currentBranch.GetLastItem()
+		if err != nil {
+			return 0, 0, nil, err
+		}
+		targetEndEventID = lastItem.GetEventID()
+		targetEndEventVersion = lastItem.GetVersion()
+	}
```

Comment:
```go
if request.StartEventId == nil || request.StartEventVersion == nil {
	firstItem, err := currentBranch.GetFirstItem()
	if err != nil {
		return 0, 0, nil, err
	}
	request.StartEventId = common.Int64Ptr(common.FirstEventID-1)
	request.StartEventVersion = common.Int64Ptr(firstItem.GetVersion())
}
if request.EndEventId == nil || request.EndEventVersion == nil {
	lastItem, err := currentBranch.GetLastItem()
	if err != nil {
		return 0, 0, nil, err
	}
	request.EndEventId = common.Int64Ptr(lastItem.GetEventID())
	request.EndEventVersion = common.Int64Ptr(lastItem.GetVersion())
}

targetStartEventID := request.GetStartEventId()
targetStartEventVersion := request.GetStartEventVersion()
targetEndEventID := request.GetEndEventId()
targetEndEventVersion := request.GetEndEventVersion()
```

Reply:
I don't want to change the fields in GetWorkflowExecutionRawHistoryV2Request

Reply:
Synced offline. Updated it as suggested